File: iOS/CA2D/World.swift
import Foundation

protocol WorldDelegate: AnyObject {
    func world(_ world: World, didChangeRule rule: Rule)
}

final class World {
    let width, height: Int
    var rule: Rule {
        didSet {
            shuffle()

            delegate?.world(self, didChangeRule: rule)
        }
    }
    var cells: [Int]
    weak var delegate: WorldDelegate?

    init(width: Int, height: Int, rule: Rule) {
        self.width = width
        self.height = height
        self.rule = rule
        self.cells = Array(repeating: 0, count: width * height)
    }

    func shuffle() {
        let rate = 0.1
        cells = (0..<(width * height)).map { _ in drand48() < rate ? rule.conditions - 1 : 0 }
    }

    func tick() {
        var nextCells: [Int] = Array(repeating: 0, count: cells.count)
        let condMax = rule.conditions - 1

        func index(_ idx: (Int, Int)) -> Int {
            let (x, y) = idx
            let adjustX = (x + width) % width
            let adjustY = (y + height) % height
            return adjustY * width + adjustX
        }

        for y in 0..<self.height {
            for x in 0..<self.width {
                let indexes = [
                    (x - 1, y - 1), (x, y - 1), (x + 1, y - 1),
                    (x - 1, y    ),             (x + 1, y    ),
                    (x - 1, y + 1), (x, y + 1), (x + 1, y + 1),
                ]
                let count = indexes.map { self.cells[index($0)] == condMax ? 1 : 0 }.reduce(0){ $0 + $1 }
                let env = 1 << count
                let idx = index((x, y))
                let prevCond = self.cells[idx]

                if prevCond == 0 && (self.rule.born & env) > 0 {
                    nextCells[idx] = condMax
                } else if prevCond == condMax && (self.rule.survive & env) > 0 {
                    nextCells[idx] = condMax
                } else {
                    if prevCond > 0 {
                        nextCells[idx] = prevCond - 1
                    }
                }
            }
        }
        self.cells = nextCells
    }
}

File: Android/app/src/main/kotlin/cc/clv/android/ca2d/World.kt
package cc.clv.android.ca2d

import java.util.*
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class World(width: Int, height: Int, rule: Rule) {
    val width = width
    val height = height
    var rule = rule
        set(value) {
            lock.withLock {
                field = value
                running = false
                shuffle()
            }
        }
    var cells = ByteArray(width * height)
    var running = false
        private set
    val lock = ReentrantLock()

    init {
        shuffle()
    }

    fun shuffle() {
        lock.withLock {
            val random = Random()
            val rate = 0.1

            var newCells = ByteArray(width * height)
            for (i in 0..(width * height - 1)) {
                val condMax = (rule.conditions - 1).toByte()
                newCells[i] = if (random.nextDouble() < rate) condMax else 0
            }
            cells = newCells
        }
    }

    fun tick() {
        lock.withLock {
            var nextCells = ByteArray(width * height)
            val condMax = (rule.conditions - 1).toByte()

            fun index(x: Int, y: Int): Int {
                val adjustX = (x + width) % width
                val adjustY = (y + height) % height

                return adjustY * width + adjustX
            }

            for (y in 0..(height - 1)) {
                for (x in 0..(width - 1)) {

                    var count = 0
                    for (t in (y - 1)..(y + 1)) {
                        for (s in (x - 1)..(x + 1)) {
                            if (s == x && t == y) {
                                continue
                            }
                            if (cells[index(s, t)] == condMax) {
                                count++
                            }
                        }
                    }

                    val env = 1 shl count
                    val idx = index(x, y)
                    val prevCond = cells[idx]

                    if (prevCond == 0.toByte() && (rule.born and env) > 0) {
                        nextCells[idx] = condMax
                    } else if (prevCond == condMax && (rule.survive and env) > 0) {
                        nextCells[idx] = condMax
                    } else {
                        if (prevCond > 0) {
                            nextCells[idx] = (prevCond - 1).toByte()
                        }
                    }
                }
            }

            cells = nextCells
        }
    }

    fun toggleRunning() {
        running = !running
    }
}
